File: LivrusVirtus/Delphi8.NET/WinForm/fontes/RotinasGlobais.pas
unit RotinasGlobais;

interface

uses System.Collections, System.Windows.Forms, System.Data, System.Data.OleDb;

type
  TRotinasGlobais = class(System.Object)
  private
    { Private Declarations }
  public
    sUsuarioLogin, sUsuarioSenha: string;
    sCodigoSelecionado, sPreco, sQtdEstoque: string;
    function ConsultaDados(Consulta: OleDbCommand; sTextoSql: string): boolean;
      overload;
    function ConsultaDados(dsConsulta: DataSet; Registros: ArrayList;
      sTextoSql: string): boolean; overload;
    function ConsultaDados(Registros: ArrayList; sTextoSql: string): boolean;
      overload;
    function ValidaCampo(sCampo, sMensagemErro: string): boolean;
    function ValidaInteger(sNum: string): boolean;
    function ValidaFloat(sNum: string): boolean;
    function VirgulaParaPonto(sNum: string; revertido: boolean): string;
    function modulo11(svalor: string; ndig: integer): integer;
    function ValidaCPF(cpf: string): boolean;
    function ValidaISBN(isbn: string): boolean;
    function ConsultaCampoDesc(Campos: ArrayList; sTextoSql,
      sMensagemErro:String): String; overload;
    function ConsultaCampoDesc(Campos: ArrayList; sCampoCodigo, sTextoSql,
      sMensagemErro: String): String; overload;
    function retDataStr(sData: DateTime; formato: string): String;
    function FormataDataStr(sData, formato: String): String;
    procedure AdicionaItemLista(lstLista: ListBox; slCodLista: ArrayList;
      sCampoCodigo: TextBox; sCampoDesc: TextBox);
    procedure RemoveItemLista(lstLista: ListBox; slCodLista: ArrayList);
    procedure AdicionaValoresLista(ConsLista: ArrayList; sTextoSql: String;
      lstLista: ListBox; slCodLista: ArrayList);

    constructor Create;
  end;

var
  Rotinas: TRotinasGlobais;

implementation

uses Principal;

constructor TRotinasGlobais.Create;
begin
  inherited Create;
  // TODO: Add any constructor code here
end;

function TRotinasGlobais.ValidaInteger(sNum: string): boolean;
begin
 if sNum<>'' then begin
    try
      Int32.Parse(sNum);
      result:=true;
    except
   	  MessageBox.Show('Número inválido!', 'Erro',
	    MessageBoxButtons.OK,MessageBoxIcon.Error);
      result:=false;
    end;
 end else
   result:=false;
end;

function TRotinasGlobais.ValidaFloat(sNum: string): boolean;
begin
 if sNum<>'' then begin
    try
      Convert.ToSingle(sNum);
      result:=true;
    except
   	  MessageBox.Show('Número inválido!', 'Erro',
	    MessageBoxButtons.OK,MessageBoxIcon.Error);
      result:=false;
    end;
 end else
   result:=false;
end;

function TRotinasGlobais.VirgulaParaPonto(sNum: string;
revertido: boolean): string;
begin
 if revertido then
   result:= sNum.Replace('.',',')
 else
   result:= sNum.Replace(',', '.');
end;

function TRotinasGlobais.modulo11(svalor: string; ndig: integer): integer;
var ncont, nsoma, nvalor, npos: integer;
    sres, slocal: string;
begin
 ncont:=1;
 while ncont <= 11 do begin
   slocal:=copy(svalor,ncont,1) + slocal;
   ncont:=ncont+1;
 end;
 nsoma:=0;
 if ndig=1 then
   ncont:=3
 else
   ncont:=2;
 npos:=2;
 while ncont <= 11 do begin
   sres:=copy(slocal,ncont,1);
   nvalor:=Int32.Parse(sres);
   nsoma:=nsoma + (nvalor * npos);
   ncont:=ncont+1;
   npos:=npos+1;
 end;
 nsoma:=(11 - (nsoma mod 11));

if nsoma > 9 then
  modulo11:=0
else
  modulo11:=nsoma;
end;

function TRotinasGlobais.ValidaCPF(cpf: string): boolean;
var scpf: string;
begin

if (length(cpf)=14) then begin
  scpf:=copy(cpf,1,3)+copy(cpf,5,3)+copy(cpf,9,3)+copy(cpf,13,2);
  if ( (modulo11(scpf,1)=Int32.parse(copy(cpf,13,1))) and
  (modulo11(scpf,2)=Int32.parse(copy(cpf,14,1))) ) then
    result:=true
  else
    result:=false;
end else
  result:=false;
end;

function TRotinasGlobais.ValidaISBN(isbn: string): boolean;
var snum, sdigito, sisbn: string;
 nvalor, nsoma, nstatus, ncont: integer;
begin
if (length(isbn)=13) then begin
  sdigito := copy(isbn, length(isbn), 1);
  sisbn:='';
  for ncont:=1 to length(isbn)-1 do begin
    if (copy(isbn,ncont,1)<>'-') then
      sisbn:=sisbn+copy(isbn,ncont,1);
  end;
  if sdigito = 'X' then
    nsoma := 10
  else begin
    nvalor:=Int32.Parse(sdigito);
    nsoma := nvalor;
  end;
  for ncont:=1 to 9 do begin
    snum:=copy(sisbn, ncont, 1);
    nvalor:=Int32.Parse(snum);
    nsoma:=nsoma + (nvalor * (11 - ncont));
  end;
  If (nsoma mod 11) = 0 Then
    validaISBN:=true
  else
    validaISBN:=false;
end else
  validaISBN:=false;
end;

function TRotinasGlobais.ConsultaDados(Consulta: OleDbCommand;
sTextoSql: string): boolean;
var nTemLinhas: boolean;
  Leitor: OleDbDataReader;
begin
  Consulta := fPrincipal.db.CreateCommand;
  Consulta.CommandText := sTextoSql;

  Leitor := Consulta.ExecuteReader;
  nTemLinhas := Leitor.HasRows;
  Leitor.Close;
  ConsultaDados := nTemLinhas;
end;

function TRotinasGlobais.ConsultaDados(dsConsulta: DataSet;
Registros: ArrayList; sTextoSql: string): boolean;
var nTemLinhas: boolean;
  Consulta: OleDbCommand;
  daConsulta: OleDbDataAdapter;
  Leitor: OleDbDataReader;
  Campos: ArrayList;
  cont: integer;
begin
  dsConsulta.Clear;
  Consulta := OleDbCommand.Create;
  daConsulta := OleDbDataAdapter.Create;

  daConsulta.TableMappings.Add('Table','Consulta');
  Consulta := fPrincipal.db.CreateCommand;
  Consulta.CommandText:=sTextoSql;
  daConsulta.SelectCommand := Consulta;
  daConsulta.Fill(dsConsulta);

  Registros.Clear;
  Leitor := Consulta.ExecuteReader;
  while Leitor.Read do begin
   Campos := ArrayList.Create;
   for cont:=0 to (Leitor.FieldCount - 1) do begin
     Campos.Add(Leitor.GetValue(cont).ToString);
   end;
   Registros.Add(Campos);
  end;
  nTemLinhas := Leitor.HasRows;
  Leitor.Close;
  ConsultaDados := nTemLinhas;
end;

function TRotinasGlobais.ConsultaDados(Registros: ArrayList;
sTextoSql: string): boolean;
var nTemLinhas: boolean;
  Consulta: OleDbCommand;
  Leitor: OleDbDataReader;
  Campos: ArrayList;
  cont: integer;
begin
  Consulta := fPrincipal.db.CreateCommand;
  Consulta.CommandText := sTextoSql;

  Registros.Clear;
  Leitor := Consulta.ExecuteReader;
  while Leitor.Read do begin
   Campos := ArrayList.Create;
   for cont:=0 to (Leitor.FieldCount - 1) do begin
     Campos.Add(Leitor.GetValue(cont).ToString);
   end;
   Registros.Add(Campos);
  end;
  nTemLinhas := Leitor.HasRows;
  Leitor.Close;
  ConsultaDados := nTemLinhas;

end;

function TRotinasGlobais.ValidaCampo(sCampo, sMensagemErro: string): boolean;
begin
result:=false;
  if sCampo='' then begin
  	MessageBox.Show(sMensagemErro, 'Erro', MessageBoxButtons.OK,
    MessageBoxIcon.Error);
    result:=true;
  end;
end;

function TRotinasGlobais.ConsultaCampoDesc(Campos: ArrayList;
sTextoSql, sMensagemErro:String): String;
var Consulta: OleDbCommand;
  Leitor: OleDbDataReader;
  cont: integer;
  nTemLinhas: boolean;
begin
  Consulta:= fPrincipal.db.CreateCommand;
  Consulta.CommandText:= sTextoSql;

  Campos.Clear;
  Leitor:=Consulta.ExecuteReader;
  if Leitor.Read then begin
	 for cont:= 0 to Leitor.FieldCount - 1 do begin
	   Campos.Add(Leitor.GetValue(cont).ToString);
	 end;
  end;
  nTemLinhas:=Leitor.HasRows;
  Leitor.Close;

  if nTemLinhas then
	 result:=Campos.Item[1].ToString
  else begin
    MessageBox.Show(sMensagemErro, 'Erro',
    MessageBoxButtons.OK,MessageBoxIcon.Error);
	  result:='';
  end;
end;

function TRotinasGlobais.ConsultaCampoDesc(Campos: ArrayList;
sCampoCodigo, sTextoSql, sMensagemErro: String): String;
begin
  if ValidaInteger(sCampoCodigo) then
    result:=ConsultaCampoDesc(Campos, sTextoSql, sMensagemErro)
  else
    result:='';
end;

function TRotinasGlobais.retDataStr(sData: DateTime; formato: string): String;
begin
  result:=sData.ToString(formato);
end;

function TRotinasGlobais.FormataDataStr(sData, formato: String): String;
var dtNasc: DateTime;
begin
  dtNasc:=DateTime.Parse(sData);
  result:=dtNasc.ToString(formato);
end;

procedure TRotinasGlobais.AdicionaItemLista(lstLista: ListBox;
slCodLista: ArrayList; sCampoCodigo: TextBox; sCampoDesc: TextBox);
var sItems: String;
begin
  if sCampoDesc.Text<>'' then begin
     sItems:=sCampoCodigo.Text + ' - ' + sCampoDesc.Text;
     if lstLista.Items.IndexOf(sItems)=-1 then begin
    	slCodLista.Add(sCampoCodigo.Text);
    	lstLista.Items.Add(sItems);
     end;
     sCampoCodigo.Clear;
     sCampoDesc.Clear;
  end;
end;

procedure TRotinasGlobais.RemoveItemLista(lstLista: ListBox;
slCodLista: ArrayList);
begin
 if lstLista.Items.Count > 0 then begin
   slCodLista.RemoveAt(lstLista.SelectedIndex);
   lstLista.Items.RemoveAt(lstLista.SelectedIndex);
 end;
end;

procedure TRotinasGlobais.AdicionaValoresLista(ConsLista: ArrayList;
sTextoSql: String; lstLista: ListBox; slCodLista: ArrayList);
var Consulta: OleDbCommand;
Leitor: OleDbDataReader;
begin
 ConsLista.Clear;
 lstLista.Items.Clear;
 slCodLista.Clear;

  Consulta:= fPrincipal.db.CreateCommand;
  Consulta.CommandText := sTextoSql;

  Leitor:= Consulta.ExecuteReader;
  while Leitor.Read do begin
    slCodLista.Add(Leitor.GetValue(0).ToString);
    lstLista.Items.Add(Leitor.GetValue(0).ToString+' - '+
    Leitor.GetValue(1).ToString);
  end;
  Leitor.Close;

end;

end.

File: LivrusVirtus/CSBuilder/CS/RotinasGlobais.cs
using System;
using System.Windows.Forms;
using System.Data;
using System.Data.OleDb;
using System.Collections;

namespace Livrus
{
	/// <summary>
	/// Summary description for Class.
	/// </summary>
	public class RotinasGlobais
	{
		public static RotinasGlobais Rotinas;
		public string sUsuarioLogin, sUsuarioSenha;
		public string sCodigoSelecionado, sPreco, sQtdEstoque;

		public RotinasGlobais()
		{
			//
			// TODO: Add constructor logic here
			//
		}

//---------------------------------------------------------------------------
public Boolean ValidaInteger(string sNum)
{
 if (sNum!="") {
	try {
	  int.Parse(sNum);
	  return true;
	} catch {
	  MessageBox.Show("Número inválido!", "Erro",
	  MessageBoxButtons.OK,MessageBoxIcon.Error);
	  return false;
	}
 } else
   return false;
}
//---------------------------------------------------------------------------
public Boolean ValidaFloat(string sNum)
{
 if (sNum!="") {
	try {
	  float.Parse(sNum);
	  return true;
	} catch {
	  MessageBox.Show("Número inválido!", "Erro",
	  MessageBoxButtons.OK,MessageBoxIcon.Error);
	  return false;
	}
 } else
   return false;
}
//---------------------------------------------------------------------------
public string VirgulaParaPonto(string sNum, Boolean revertido)
{
 if (revertido)
   return sNum.Replace(".",",");
 else
   return sNum.Replace(",",".");
}
//---------------------------------------------------------------------------
public int modulo11(string svalor, int ndig)
{
int ncont, nsoma, nvalor, npos;
string slocal, sres;

 ncont=0;
 slocal="";
 while (ncont < 11) {
   slocal=svalor.Substring(ncont,1) + slocal;
   ncont++;
 }
 nsoma=0;
 if (ndig == 1)
   ncont=2;
 else
   ncont=1;
 npos=2;
 while (ncont < 11) {
   sres = slocal.Substring(ncont,1);
   nvalor = int.Parse(sres);
   nsoma=nsoma + (nvalor * npos);
   ncont++;
   npos++;
 }
 nsoma=(11 - (nsoma % 11));

 if (nsoma > 9)
   return 0;
 else
   return nsoma;
}
//---------------------------------------------------------------------------
public Boolean ValidaCPF(string cpf)
{
string scpf;

if (cpf.Length==14) {
  scpf=cpf.Substring(0,3)+cpf.Substring(4,3)+cpf.Substring(8,3)+
	cpf.Substring(12,2);
  if ( (modulo11(scpf,1)==int.Parse(cpf.Substring(12,1))) &&
  (modulo11(scpf,2)==int.Parse(cpf.Substring(13,1))) )
	return true;
  else
	return false;
} else
  return false;
}
//---------------------------------------------------------------------------
public Boolean ValidaISBN(string isbn)
{
 string sisbn, sdigito, snum;
 int nvalor, nsoma, ncont;

if (isbn.Length==13) {
  sdigito = isbn.Substring(isbn.Length-1, 1);
  sisbn="";
  for(ncont = 0; ncont < isbn.Length-1; ncont++)  {
    if (isbn.Substring(ncont,1)!="-")
      sisbn = sisbn + isbn.Substring(ncont,1);
  }
  if (sdigito=="X")
    nsoma = 10;
  else {
    nvalor = int.Parse(sdigito);
    nsoma = nvalor;
  }
  for(ncont = 1; ncont <= 9; ncont++)  {
    snum = sisbn.Substring(ncont-1, 1);
    nvalor = int.Parse(snum);
    nsoma=nsoma + (nvalor * (11 - ncont));
  }
  if ((nsoma % 11) == 0)
    return true;
  else
    return false;
} else
  return false;
}
//---------------------------------------------------------------------------
public bool ConsultaDados(OleDbCommand Consulta, string sTextoSql)
{
bool nTemLinhas = false;

  Consulta = frmPrincipal.fPrincipal.db.CreateCommand();
  Consulta.CommandText = sTextoSql;

  OleDbDataReader Leitor = Consulta.ExecuteReader();
  nTemLinhas = Leitor.HasRows;
  Leitor.Close();
  return nTemLinhas;

}
//---------------------------------------------------------------------------
public bool ConsultaDados(DataSet dsConsulta, ArrayList Registros,
						  string sTextoSql)
{
bool nTemLinhas = false;

  dsConsulta.Clear();
  OleDbCommand Consulta = new OleDbCommand();
  OleDbDataAdapter daConsulta = new OleDbDataAdapter();

  daConsulta.TableMappings.Add("Table","Consulta");
  Consulta = frmPrincipal.fPrincipal.db.CreateCommand();
  Consulta.CommandText=sTextoSql;
  daConsulta.SelectCommand = Consulta;
  daConsulta.Fill(dsConsulta);

  Registros.Clear();
  OleDbDataReader Leitor = Consulta.ExecuteReader();
  while (Leitor.Read()) {
	 ArrayList Campos = new ArrayList();
	 for(int cont=0; cont < Leitor.FieldCount; cont++) {
	   Campos.Add(Leitor.GetValue(cont).ToString());
	 }
	 Registros.Add(Campos);
  }
  nTemLinhas = Leitor.HasRows;
  Leitor.Close();
  return nTemLinhas;
}
//---------------------------------------------------------------------------
public bool ConsultaDados(ArrayList Registros, string sTextoSql)
{
bool nTemLinhas = false;

  OleDbCommand Consulta = frmPrincipal.fPrincipal.db.CreateCommand();
  Consulta.CommandText = sTextoSql;

  Registros.Clear();
  OleDbDataReader Leitor = Consulta.ExecuteReader();
  while (Leitor.Read()) {
	 ArrayList Campos = new ArrayList();
	 for(int cont=0; cont < Leitor.FieldCount; cont++) {
	   Campos.Add(Leitor.GetValue(cont).ToString());
	 }
	 Registros.Add(Campos);
  }
  nTemLinhas = Leitor.HasRows;
  Leitor.Close();
  return nTemLinhas;

}
//---------------------------------------------------------------------------
public Boolean ValidaCampo(string sCampo,string sMensagemErro)
{
  if (sCampo=="") {
	MessageBox.Show(sMensagemErro, "Erro",
	MessageBoxButtons.OK,MessageBoxIcon.Error);
	return true;
  } else
	return false;
}
//---------------------------------------------------------------------------
public string ConsultaCampoDesc(ArrayList Campos,
				string sTextoSql, string sMensagemErro)
{

  OleDbCommand Consulta = frmPrincipal.fPrincipal.db.CreateCommand();
  Consulta.CommandText = sTextoSql;

  Campos.Clear();
  OleDbDataReader Leitor = Consulta.ExecuteReader();
  if (Leitor.Read()) {
	 for(int cont=0; cont < Leitor.FieldCount; cont++) {
	   Campos.Add(Leitor.GetValue(cont).ToString());
	 }
  }
  bool nTemLinhas = Leitor.HasRows;
  Leitor.Close();

  if (nTemLinhas)
	 return Campos[1].ToString();
  else {
	MessageBox.Show(sMensagemErro, "Erro",
	MessageBoxButtons.OK,MessageBoxIcon.Error);
	return "";
  }
}
//---------------------------------------------------------------------------
public string ConsultaCampoDesc(ArrayList Campos,
		  string sCampoCodigo,string sTextoSql,string sMensagemErro)
{
if (ValidaInteger(sCampoCodigo))
  return ConsultaCampoDesc(Campos, sTextoSql, sMensagemErro);
else
  return "";
}
//---------------------------------------------------------------------------
  public string retDataStr(DateTime sData, string formato) {
	return sData.ToString(formato);
  }
//---------------------------------------------------------------------------
  public string FormataDataStr(string sData, string formato) {
	DateTime dtNasc = DateTime.Parse(sData);
	return dtNasc.ToString(formato);
  }

//---------------------------------------------------------------------------
public void AdicionaItemLista(ListBox lstLista,
	 ArrayList slCodLista, TextBox sCampoCodigo, TextBox sCampoDesc)
{
string sItems;

if (sCampoDesc.Text.Equals("")==false) {
 sItems=sCampoCodigo.Text+" - "+sCampoDesc.Text;
 if (lstLista.Items.IndexOf(sItems)==-1) {
	slCodLista.Add(sCampoCodigo.Text);
	lstLista.Items.Add(sItems);
 }
 sCampoCodigo.Clear();
 sCampoDesc.Clear();
}
}
//---------------------------------------------------------------------------
public void RemoveItemLista(ListBox lstLista, ArrayList slCodLista)
{
 if (lstLista.Items.Count > 0) {
   slCodLista.RemoveAt(lstLista.SelectedIndex);
   lstLista.Items.RemoveAt(lstLista.SelectedIndex);
 }
}
//---------------------------------------------------------------------------
public void AdicionaValoresLista(ArrayList ConsLista,
	 string sTextoSql,ListBox lstLista, ArrayList slCodLista)
{
 ConsLista.Clear();
 lstLista.Items.Clear();
 slCodLista.Clear();

  OleDbCommand Consulta = frmPrincipal.fPrincipal.db.CreateCommand();
  Consulta.CommandText = sTextoSql;

  OleDbDataReader Leitor = Consulta.ExecuteReader();
  while (Leitor.Read()) {
	  slCodLista.Add(Leitor.GetValue(0).ToString());
	  lstLista.Items.Add(Leitor.GetValue(0).ToString()+" - "+
	  Leitor.GetValue(1).ToString());
  }
  Leitor.Close();

}
//---------------------------------------------------------------------------
}
}
